src/computability/tm_to_partrec.lean

```diff
@@ -0,0 +1,885 @@
+    intros a b e, induction b with b IH generalizing a e,
+    { refine pfun.mem_fix_iff.2 (or.inl $ roption.eq_some_iff.1 _),
+      simp only [hg, ← e, pure_bind, list.tail_cons], refl },
+    { refine pfun.mem_fix_iff.2 (or.inr ⟨_, _, IH (a+1) (by rwa add_right_comm)⟩),
+      simp only [hg, eval, pure_bind, nat.elim_succ, list.tail],
+      exact roption.mem_some_iff.2 rfl } },
+  case nat.partrec'.comp : m n f g hf hg IHf IHg { exact exists_code.comp IHf IHg },
+  case nat.partrec'.rfind : n f hf IHf {
+    obtain ⟨cf, hf⟩ := IHf, refine ⟨rfind cf, λ v, _⟩,
+    replace hf := λ a, hf (a :: v),
+    simp only [roption.map_eq_map, roption.map_some, vector.cons_val, pfun.coe_val] at hf ⊢,
+    refine roption.ext (λ x, _),
+    simp only [rfind, roption.bind_eq_bind, roption.pure_eq_some, roption.map_eq_map,
+      roption.bind_some, exists_prop, eval, list.head, pred_eval, roption.map_some,
+      bool.ff_eq_to_bool_iff, roption.mem_bind_iff, list.length,
+      roption.mem_map_iff, nat.mem_rfind, list.tail, bool.tt_eq_to_bool_iff,
+      roption.mem_some_iff, roption.map_bind],
+    split,
+    { rintro ⟨v', h1, rfl⟩,
+      suffices : ∀ (v₁ : list ℕ), v' ∈ pfun.fix
+        (λ v, (cf.eval v).bind $ λ y, roption.some $ if y.head = 0 then
+          sum.inl (v.head.succ :: v.tail) else sum.inr (v.head.succ :: v.tail)) v₁ →
+        ∀ n, v₁ = n :: v.val → (∀ m < n, ¬f (m :: v) = 0) →
+        (∃ (a : ℕ), (f (a :: v) = 0 ∧ ∀ {m : ℕ}, m < a → ¬f (m :: v) = 0) ∧ [a] = [v'.head.pred]),
+      { exact this _ h1 0 rfl (by rintro _ ⟨⟩) },
+      clear h1, intros v₀ h1,
+      refine pfun.fix_induction h1 (λ v₁ h2 IH, _), clear h1,
+      rintro n rfl hm,
+      have := pfun.mem_fix_iff.1 h2,
+      simp only [hf, roption.bind_some] at this,
+      split_ifs at this,
+      { simp only [list.head, exists_false, or_false, roption.mem_some_iff,
+          list.tail_cons, false_and] at this,
+        subst this, exact ⟨_, ⟨h, hm⟩, rfl⟩ },
+      { simp only [list.head, exists_eq_left, roption.mem_some_iff,
+          list.tail_cons, false_or] at this,
+        refine IH _ this (by simp [hf, h]) _ rfl (λ m h', _),
+        obtain h|rfl := nat.lt_succ_iff_lt_or_eq.1 h', exacts [hm _ h, h] } },
+    { rintro ⟨n, ⟨hn, hm⟩, rfl⟩, refine ⟨n.succ :: v.1, _, rfl⟩,
+      have : (n.succ :: v.1 : list ℕ) ∈ pfun.fix
+        (λ v, (cf.eval v).bind $ λ y, roption.some $ if y.head = 0 then
+          sum.inl (v.head.succ :: v.tail) else sum.inr (v.head.succ :: v.tail)) (n :: v.val) :=
+        pfun.mem_fix_iff.2 (or.inl (by simp [hf, hn])),
+      generalize_hyp : (n.succ :: v.1 : list ℕ) = w at this ⊢, clear hn,
+      induction n with n IH, {exact this},
+      refine IH (λ m h', hm (nat.lt_succ_of_lt h')) (pfun.mem_fix_iff.2 (or.inr ⟨_, _, this⟩)),
+      simp only [hf, hm n.lt_succ_self, roption.bind_some, list.head, eq_self_iff_true,
+        if_false, roption.mem_some_iff, and_self, list.tail_cons] } }
+end
+
+end code
+
+inductive cont
```

Comment:
I think this deserves some explanation.  Is this some kind of Krivine machine?

Reply:
I'm in the middle of a big docs push, I should have something in a few hours at most, but I believe it is quite similar to a Krivine machine (except it's not evaluating lambda terms). It's also call by value, which matches the monadic definition of `code.eval` and eval functions from partial recursive functions pretty well.